Added Pool3DDNNLayer and MaxPool3DDNNLayer

gyglim opened the pull request:
Hi

This adds the cuDnn implementation of the Pool3DDNNLayer. Required by https://github.com/Lasagne/Recipes/pull/41.

Cheers,
Michael

f0k commented on lasagne/layers/dnn.py:
```diff
@@ -16,6 +16,8 @@
     "Pool2DDNNLayer",
     "MaxPool2DDNNLayer",
+    "Pool3DDNNLayer",
+    "MaxPool3DDNNLayer",
```
Please also add this to `docs/modules/layers.rst`!

f0k commented on lasagne/layers/dnn.py:
```diff
@@ -125,6 +127,114 @@ def __init__(self, incoming, pool_size, stride=None,
+    def __init__(self, incoming, pool_size, stride=None, pad=(0, 0, 0),
+                 ignore_border=True, mode='max', **kwargs):
+        super(Pool3DDNNLayer, self).__init__(incoming, **kwargs)
+        self.pool_size = as_tuple(pool_size, 3)
```
It was merged into `master` just minutes ago, but could you please also copy/paste the dimensionality check of #569 here?

f0k replied:
(It will probably be easier if you rebase your PR onto the latest master, so the existing 2d pooling tests will include the dimensionality check. Let me know if you need help.)

f0k commented on lasagne/layers/dnn.py:
```diff
@@ -125,6 +127,114 @@ def __init__(self, incoming, pool_size, stride=None,
+                                             stride=self.stride[1],
+                                             pad=self.pad[1],
+                                             ignore_border=True,
+                                             )
```
Nitpick: Just for consistency, please also add a newline here.

f0k commented on lasagne/tests/layers/test_pool.py:
```diff
@@ -502,6 +523,88 @@ def test_not_implemented(self):
+        except ImportError:
+            pytest.skip("cuDNN not available")
+        with pytest.raises(NotImplementedError) as exc:
+            layer = MaxPool3DDNNLayer((1, 2, 3, 4), pool_size=2,
```
Should be `(1, 2, 3, 4, 5)` (but you'd have noticed after adding the dimensionality check).

f0k commented:
Thank you, looks good except for a few minor things! If you can fix them and squash your PR into a single commit, this is good to merge. (If you need any help regarding rebasing/squashing, just let me know.) Cheers!

gyglim commented:
Great, thank. I made the changes. 
On a related note: why is the indexing done with [0],[1] and so on for the Pooling? Could be done from the back, no ([-1], [-2] and so on)? This would remove the constraint that the tensor needs to have exactly 5 (or 4 in the 2D pooling case) dimensions, as long as there are as many as there are pooled over?

As for the squashing: Yes, I've never done it and would appreciate some help

f0k commented:
> This would remove the constraint that the tensor needs to have exactly 5 (or 4 in the 2D pooling case) dimensions, as long as there are as many as there are pooled over?

It would be easy for us to change the implementation in `get_output_shape_for()`, but the cuDNN implementation requires 4D or 5D tensors. We'd need to include some reshaping in `get_output_for()` then to squash the first dimensions before `dnn_pool()` and get them back out afterwards (if the input dimensionality is too large), or to introduce additional first dimensions and remove them later (if the input dimensionality is too low). As long as there are not obvious use cases for this, I'd leave this burden up to the user.

> Great, thank. I made the changes. 

Great! You'll also need to add a test for the dimensionality check, like this one: https://github.com/Lasagne/Lasagne/pull/569/files#diff-05ecd45be63aafa8322cac30a5c0c66eR522
Just adapt and add it to the end of `TestMaxPool3DNNLayer`. You're lucky that Travis doesn't have a GPU, otherwise coveralls would have complained.

> As for the squashing: Yes, I've never done it and would appreciate some help

The easiest way to remember is to run `git rebase --interactive master`. You'll be greeted with a text editor listing all your commits and giving some guidance below. You'll have to leave the first commit as it is (unless you want to reword it, using `r`). All the other commits can be changed to `f` (for fixup). When you save and exit, git will squash all `fixup` commits into the first one.

f0k commented:
It seems you pulled in a commit of @danstowell into your PR now? This creates a merge conflict... you can go back with `git reset --hard 1ab40a4; git push --force`. To get the upstream changes into your PR, it's better to rebase it. Use `git fetch upstream; git rebase upstream/master` to rebase your PR onto the latest master (assuming you've got the official Lasagne repository named `upstream` in your clone).

gyglim commented:
Ok, I added the ndim test. 
In the test code from https://github.com/Lasagne/Lasagne/pull/569/files there was a small error, it used  MaxPool2DNNLayer instead of  MaxPool2DDNNLayer. Fixed this aswell. Now I just need to get the squashing right!

f0k commented on lasagne/tests/layers/test_pool.py:
```diff
@@ -525,13 +546,107 @@ def test_fail_on_mismatching_dimensionality(self):
         assert "Expected 4 input dimensions" in exc.value.args[0]
         with pytest.raises(ValueError) as exc:
-            MaxPool2DNNLayer((10, 20, 30, 40, 50), 3, 2)
+            MaxPool2DDNNLayer((10, 20, 30, 40, 50), 3, 2)
```
Oops, good catch. Forgot that we can't rely on Travis for any tests involving the GPU. I'll remember to test everything locally next time.

gyglim commented:
Hi. Ok, now I believe it is all correct, right?

f0k commented:
> Ok, now I believe it is all correct, right?

Looks good, except that the newline (https://github.com/Lasagne/Lasagne/pull/572#discussion_r49480283) seems to have disappeared again? Didn't you fix this already?

After fixing this, do `git add lasagne/layers/dnn.py; git commit --amend --no-edit; git push --force` to integrate the change into the current commit (so you don't have to squash again).

gyglim commented:
I added one, but forgot another. Fixed it now

f0k commented:
Looks good to me! Tests pass on my machine. Will merge tomorrow if there are no further comments.

gyglim commented:
Great, thank you for guiding me through the process

benanne commented:
Awesome, thanks a lot for contributing!

f0k commented:
Merging, thanks again!
